src/flow/domain/mod.rs

```diff
@@ -1502,6 +1505,21 @@ impl Domain {
         }
     }
 
+    fn handle_recovery(&mut self) {
+        let packets = persistence::retrieve_recovery_packets(
+            &self.nodes,
+            self.index,
+            self.shard.unwrap_or(0),
+            &self.persistence_parameters,
+            self.transaction_state.get_checktable().clone(),
+        );
+
+        packets.into_iter().for_each(|packet| self.handle(packet));
```

Comment:
It is kind of wasteful to construct a Vec if you are going to immediately call into_iter on it. Instead, you could just return an iterator of all the packets.

Reply:
Agreed - but I couldn't find a clean way of returning an iterator because of the borrow on `node`. Tried something like this:

```rust
    nodes
        .iter()
        .filter_map(|(_index, node)| {
            let node = node.borrow();
            let local_addr = node.local_addr();
            let global_addr = node.global_addr();
            let path = params.log_path(&local_addr, domain_index, domain_shard);
            let file = match File::open(&path) {
                Ok(f) => f,
                Err(ref e) if e.kind() == ErrorKind::NotFound => return None,
                Err(e) => panic!("Could not open log file {:?}: {}", path, e),
            };

            let iter = BufReader::new(file)
                .lines()
                .filter_map(|line| {
                    let line = line.unwrap();
                    let entries: Result<Vec<Records>, _> = serde_json::from_str(&line);
                    entries.ok()
                })
                .flat_map(|r| r)
                // Merge packets into batches of RECOVERY_BATCH_SIZE:
                .chunks(RECOVERY_BATCH_SIZE)
                .into_iter()
                .map(|chunk| chunk.fold(Records::default(), |mut acc, ref mut data| {
                    acc.append(data);
                    acc
                }))
                // Then create Packet objects from the data:
                .map(|data| {
                    let link = Link::new(*local_addr, *local_addr);
                    if node.is_transactional() {
                        let (ts, prevs) = checktable
                            .recover(global_addr, node.fields().len())
                            .unwrap();

                        box Packet::Transaction {
                            link,
                            data,
                            tracer: None,
                            state: TransactionState::Committed(ts, global_addr, prevs),
                        }
                    } else {
                        box Packet::Message {
                            link,
                            data,
                            tracer: None,
                        }
                    }
                });

            Some(iter)
        })
        .flat_map(|i| i)
```

Reply:
I see the issue. Could something like the following work?

```rust
fn handle_recovery(&mut self) {
    let addresses: Vec<_> = nodes.iter().map(|n| (n.local_addr(), n.global_addr())).collect();
    for (local, global) in addresses {
         for packet in persistence::retrieve_recovery_packets(...) {
              self.handle(packet);
         }
    }
}
```

Reply:
@fintelia: I ended up moving the `retrieve_recovery_packets` logic up to `handle_recovery` - some of the arguments ended up looking a bit weird, so not sure if it was a very good abstraction anyway. Let me know if you disagree though and I'll take a second look at it.

For reference I also bumped into a separate issue when I tried returning an iterator from `retrieve_recovery_packets`: `.chunks()` returns `IntoChunks` - [not an iterable](https://docs.rs/itertools/0.7.2/itertools/trait.Itertools.html#method.chunks) which meant that the value [here](https://gist.github.com/ekmartin/03569ec5f2d8bde35efcac918d71ef44#file-persistence-rs-L17) didn't last as long as `'a`.